File: packages/sdk-js/src/v2/live/session.ts
import { EventEmitter } from 'eventemitter3'
import { concatArrayBuffer, toUint8Array } from '../../helpers.js'
import { HttpClient } from '../../network/httpClient.js'
import { WebSocketClient, WebSocketSession, WS_STATES } from '../../network/wsClient.js'
import type {
  LiveV2InitRequest,
  LiveV2InitResponse,
  LiveV2StartSessionMessage,
  LiveV2WebSocketMessage,
} from './generated-types.js'
import { LiveV2SessionStatus } from './types.js'

type EventMap = {
  started: [message: LiveV2InitResponse]
  connecting: [message: { attempt: number }]
  connected: [message: { attempt: number }]
  ending: [message: { code: number; reason?: string }]
  ended: [message: { code: number; reason?: string }]
  message: [message: LiveV2WebSocketMessage]
  error: [error: Error]
}

export class LiveV2Session {
  private sessionOptions: LiveV2InitRequest
  private httpClient: HttpClient
  private webSocketClient: WebSocketClient

  private abortController: AbortController = new AbortController()
  private initSessionPromise: Promise<LiveV2InitResponse>
  private initSession: LiveV2InitResponse | null = null
  private webSocketSession: WebSocketSession | null = null

  private eventEmitter: EventEmitter | null = new EventEmitter()
  private bytesSent = 0
  private audioBuffer: Uint8Array | null = null

  private _status: LiveV2SessionStatus = 'starting'

  constructor({
    options,
    httpClient,
    webSocketClient,
  }: {
    options: LiveV2InitRequest
    httpClient: HttpClient
    webSocketClient: WebSocketClient
  }) {
    this.sessionOptions = options
    this.httpClient = httpClient
    this.webSocketClient = webSocketClient
    this.abortController = new AbortController()

    this.initSessionPromise = this.startSession()
    this.initSessionPromise.then((session) => {
      this.initSession = session

      if (this.abortController.signal.aborted) {
        return
      }

      if (this._status === 'starting') {
        this._status = 'started'
        this.emit('started', session)
      }

      if (this.sessionOptions.messages_config?.receive_lifecycle_events) {
        const startSessionMessage: LiveV2StartSessionMessage = {
          type: 'start_session',
          session_id: session.id,
          created_at: session.created_at,
        }
        this.emit('message', startSessionMessage)
      }

      return this.connectToWebSocket(session)
    })
  }

  /**
   * Get the session id. The promise is resolved when the session is started.
   * @returns the session id
   */
  async getSessionId(): Promise<string> {
    const { id } = await this.initSessionPromise
    return id
  }

  /**
   * The session id or null if the session is not started yet.
   */
  get sessionId(): string | null {
    return this.initSession?.id ?? null
  }

  /**
   * The current status of the session.
   * - `starting`: the session is not started yet
   * - `started`: the session is started but not connected to the websocket
   * - `connecting`: the session is connecting to the websocket. If the connection is lost and it's retryable, the session will reconnect and the status will be `connecting` again.
   * - `connected`: the session is connected to the websocket.
   * - `ending`: the session is ending because of a user action or an error. In this status, sendAudio and stop are no-op.
   * - `ended`: the session is ended. In this status, sendAudio and stop are no-op and listeners are removed.
   */
  get status(): LiveV2SessionStatus {
    return this._status
  }

  /**
   * Send an audio chunk to the current live session.
   * If not connected, the audio will be buffered and sent when the session is connected.
   *
   * @param audio the audio chunk to send
   */
  sendAudio(audio: ArrayBufferLike | Buffer<ArrayBufferLike> | ArrayLike<number>): void {
    if (this._status === 'ended') {
      // throw new Error(`The session stopped, you can no longer send audio.`)
      return
    }
    if (this._status === 'ending') {
      // throw new Error(`The session is stopping, you can no longer send audio.`)
      return
    }

    // TODO check if it has a wav header.
    // If it does, check with config and remove it

    const audioArray = toUint8Array(audio)
    this.audioBuffer = concatArrayBuffer(this.audioBuffer, audioArray)

    if (this.webSocketSession?.readyState === WS_STATES.OPEN) {
      this.webSocketSession?.send(audioArray)
    }
  }

  /**
   * Stop the recording.
   * The session will process the remaining audio and emit the `ended` event when the post-processing is done.
   */
  stopRecording(): void {
    this.doStop(1000)
  }

  /**
   * Force the end of the session without waiting for the post-processing to be done.
   * `ending` and `ended` events are emitted, pending requests/connections are cancelled and listeners are removed.
   */
  endSession(): void {
    this.doDestroy(1000, 'Session ended by user')
  }

  private doStop(code = 1006, reason?: string): void {
    if (this._status === 'ended') {
      // no-op
      return
    }
    if (this._status === 'ending') {
      // no-op
      return
    }

    this._status = 'ending'
    this.emit('ending', { code, reason })

    if (this.webSocketSession?.readyState === WS_STATES.OPEN) {
      this.webSocketSession?.send(JSON.stringify({ type: 'stop_recording' }))
    }
  }

  private doDestroy(code = 1006, reason?: string) {
    if (this._status === 'ended') {
      return
    }

    // Ending the session
    this.doStop(code, reason)

    // Session ended
    this._status = 'ended'
    this.emit('ended', { code, reason })

    // Cancel pending connection
    this.abortController.abort()

    this.audioBuffer = null

    this.removeAllListeners()
    this.eventEmitter = null
  }

  private connectToWebSocket(session: LiveV2InitResponse): void {
    if (this.abortController.signal.aborted) {
      return
    }

    // Create a WebSocket session and bridge its events
    const webSocketSession = this.webSocketClient.createSession(session.url)
    this.abortController.signal.addEventListener('abort', () => {
      this.webSocketSession = null
      webSocketSession.onconnecting = null
      webSocketSession.onopen = null
      webSocketSession.onmessage = null
      webSocketSession.onclose = null
      webSocketSession.onerror = null
      webSocketSession.close(1001, 'Aborted')
    })

    this.webSocketSession = webSocketSession

    this.webSocketSession.onconnecting = ({ connection }) => {
      this._status = 'connecting'
      this.emit('connecting', { attempt: connection })
    }

    this.webSocketSession.onopen = ({ connection }) => {
      if (this.audioBuffer?.byteLength) {
        webSocketSession.send(this.audioBuffer)
      }
      if (this.status === 'ending') {
        webSocketSession.send(JSON.stringify({ type: 'stop_recording' }))
        return
      }
      this._status = 'connected'
      this.emit('connected', { attempt: connection })
    }

    this.webSocketSession.onmessage = (event) => {
      const message = this.parseMessage(event.data.toString())

      // We forced the acknowledgment reception for resume so we must not emit them if user don't want them
      if (
        this.sessionOptions.messages_config?.receive_acknowledgments ||
        !('acknowledged' in message)
      ) {
        this.emit('message', message)
      }

      if (message.type === 'audio_chunk') {
        if (message.acknowledged && message.data) {
          const byteEnd = message.data.byte_range[1]
          const slice = this.audioBuffer?.slice(byteEnd - this.bytesSent)
          this.audioBuffer = slice?.byteLength ? slice : null
          this.bytesSent = byteEnd
        }
      }
    }

    this.webSocketSession.onclose = ({ code, reason }) => {
      this.doDestroy(code, reason)
    }

    this.webSocketSession.onerror = (error) => {
      this.emit('error', error)
    }
  }

  private async startSession(): Promise<LiveV2InitResponse> {
    try {
      return await this.httpClient.post<LiveV2InitResponse>(`/v2/live`, {
        signal: this.abortController.signal,
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify({
          ...this.sessionOptions,
          messages_config: {
            ...this.sessionOptions.messages_config,
            // Force ack reception for resume
            receive_acknowledgments: true,
          },
        } satisfies LiveV2InitRequest),
      })
    } catch (error) {
      if (!this.abortController.signal.aborted) {
        this.emit(
          'error',
          error instanceof Error
            ? error
            : new Error(`Error creating session: ${error}`, { cause: error })
        )
        this.doDestroy(1006, `Couldn't start a new session`)
      }
      throw error
    }
  }

  private parseMessage(data: string): LiveV2WebSocketMessage {
    let message: unknown
    try {
      message = JSON.parse(data)
    } catch (err) {
      throw new Error(`Invalid message received: ${data}`, { cause: err })
    }
    if (!message || typeof message !== 'object' || !('type' in message)) {
      throw new Error(`Invalid message received: ${data}`)
    }
    return message as LiveV2WebSocketMessage
  }

  // #### Listeners ####

  on(type: 'started', cb: (...args: EventMap['started']) => void): this
  on(type: 'connecting', cb: (...args: EventMap['connecting']) => void): this
  on(type: 'connected', cb: (...args: EventMap['connected']) => void): this
  on(type: 'ending', cb: (...args: EventMap['ending']) => void): this
  on(type: 'ended', cb: (...args: EventMap['ended']) => void): this
  on(type: 'message', cb: (...args: EventMap['message']) => void): this
  on(type: 'error', cb: (...args: EventMap['error']) => void): this
  // eslint-disable-next-line @typescript-eslint/no-explicit-any -- overload, cannot avoid any
  on(type: keyof EventMap, cb: (event: any) => void): this {
    this.eventEmitter?.on(type, cb)
    return this
  }

  once(type: 'started', cb: (...args: EventMap['started']) => void): this
  once(type: 'connecting', cb: (...args: EventMap['connecting']) => void): this
  once(type: 'connected', cb: (...args: EventMap['connected']) => void): this
  once(type: 'ending', cb: (...args: EventMap['ending']) => void): this
  once(type: 'ended', cb: (...args: EventMap['ended']) => void): this
  once(type: 'message', cb: (...args: EventMap['message']) => void): this
  once(type: 'error', cb: (...args: EventMap['error']) => void): this
  // eslint-disable-next-line @typescript-eslint/no-explicit-any -- overload, cannot avoid any
  once(type: keyof EventMap, cb: (event: any) => void): this {
    this.eventEmitter?.once(type, cb)
    return this
  }

  off(type: 'started', cb?: (...args: EventMap['started']) => void): this
  off(type: 'connecting', cb?: (...args: EventMap['connecting']) => void): this
  off(type: 'connected', cb?: (...args: EventMap['connected']) => void): this
  off(type: 'ending', cb?: (...args: EventMap['ending']) => void): this
  off(type: 'ended', cb?: (...args: EventMap['ended']) => void): this
  off(type: 'message', cb?: (...args: EventMap['message']) => void): this
  off(type: 'error', cb?: (...args: EventMap['error']) => void): this
  // eslint-disable-next-line @typescript-eslint/no-explicit-any -- overload, cannot avoid any
  off(type: keyof EventMap, cb?: (event: any) => void): this {
    this.eventEmitter?.off(type, cb)
    return this
  }

  addListener(type: 'started', cb: (...args: EventMap['started']) => void): this
  addListener(type: 'connecting', cb: (...args: EventMap['connecting']) => void): this
  addListener(type: 'connected', cb: (...args: EventMap['connected']) => void): this
  addListener(type: 'ending', cb: (...args: EventMap['ending']) => void): this
  addListener(type: 'ended', cb: (...args: EventMap['ended']) => void): this
  addListener(type: 'message', cb: (...args: EventMap['message']) => void): this
  addListener(type: 'error', cb: (...args: EventMap['error']) => void): this
  addListener(
    type: keyof EventMap,
    // eslint-disable-next-line @typescript-eslint/no-explicit-any -- overload, cannot avoid any
    cb: (event: any) => void
  ): this {
    this.eventEmitter?.addListener(type, cb)
    return this
  }

  removeListener(type: 'started', cb?: (...args: EventMap['started']) => void): this
  removeListener(type: 'connecting', cb?: (...args: EventMap['connecting']) => void): this
  removeListener(type: 'connected', cb?: (...args: EventMap['connected']) => void): this
  removeListener(type: 'ending', cb?: (...args: EventMap['ending']) => void): this
  removeListener(type: 'ended', cb?: (...args: EventMap['ended']) => void): this
  removeListener(type: 'message', cb?: (...args: EventMap['message']) => void): this
  removeListener(type: 'error', cb?: (...args: EventMap['error']) => void): this
  removeListener(
    type: keyof EventMap,
    // eslint-disable-next-line @typescript-eslint/no-explicit-any -- overload, cannot avoid any
    cb?: (event: any) => void
  ): this {
    this.eventEmitter?.removeListener(type, cb)
    return this
  }

  removeAllListeners(): this {
    this.eventEmitter?.removeAllListeners()
    return this
  }

  private emit(type: 'started', ...args: EventMap['started']): void
  private emit(type: 'connecting', ...args: EventMap['connecting']): void
  private emit(type: 'connected', ...args: EventMap['connected']): void
  private emit(type: 'ending', ...args: EventMap['ending']): void
  private emit(type: 'ended', ...args: EventMap['ended']): void
  private emit(type: 'message', ...args: EventMap['message']): void
  private emit(type: 'error', ...args: EventMap['error']): void
  private emit(
    type: keyof EventMap,
    // eslint-disable-next-line @typescript-eslint/no-explicit-any -- overload, cannot avoid any
    ...args: any[]
  ): void {
    this.eventEmitter?.emit(type, ...args)
  }
}

File: packages/sdk-python/src/gladiaio_sdk/v2/live/session.py
"""live v2 session implementation."""

from __future__ import annotations

import contextlib
import json
import threading
import time
from typing import Any, final

from pyee import EventEmitter

from gladiaio_sdk.v2.live.types import (
  LiveV2ConnectedMessage,
  LiveV2ConnectingMessage,
  LiveV2EndingMessage,
  LiveV2SessionStatus,
)

from ...network import (
  WS_STATES,
  HttpClient,
  WebSocketClient,
  WebSocketSession,
)
from ._helpers import (
  LiveV2SessionEventsMixin,
  emit_session_ending_events,
  emit_started_if_needed,
  maybe_emit_start_session_message,
  parse_ws_message,
  send_audio_in_chunks,
  should_emit_ws_message,
  trim_acknowledged_audio_buffer,
  with_acknowledgments_enabled,
)
from .generated_types import (
  LiveV2InitRequest,
  LiveV2InitResponse,
)


@final
class LiveV2Session(LiveV2SessionEventsMixin):
  """Live V2 session.

  Events:
  - started(LiveV2InitResponse)
  - connecting(LiveV2ConnectingMessage)
  - connected(LiveV2ConnectedMessage)
  - ending(LiveV2EndingMessage)
  - ended(LiveV2EndedMessage)
  - message(LiveV2WebSocketMessage)
  - error(Exception)
  """

  def __init__(
    self,
    *,
    options: LiveV2InitRequest,
    http_client: HttpClient,
    ws_client: WebSocketClient,
    existing_session: LiveV2InitResponse | None = None,
  ) -> None:
    self._options = options
    self._http_client = http_client
    self._ws_client = ws_client
    self._existing_session = existing_session

    self._event_emitter = EventEmitter()
    self._status: LiveV2SessionStatus = "starting"
    self._init_session_response: LiveV2InitResponse | None = None

    self._ws: WebSocketSession | None = None

    self._audio_buffer: bytes = bytes([])
    self._bytes_sent = 0
    self._state_lock = threading.Lock()
    self._ws_stop = threading.Event()
    self._ready_event = threading.Event()
    self._ws_thread: threading.Thread | None = None
    self._pending_stop = False

    # Start lifecycle in background thread (HTTP init -> WS connect -> pump)
    self._ws_thread = threading.Thread(
      target=self._lifecycle_worker,
      name="live-v2-ws",
      daemon=True,
    )
    self._ws_thread.start()

  @property
  def session_id(self) -> str | None:
    return self._init_session_response.id if self._init_session_response else None

  @property
  def status(self) -> LiveV2SessionStatus:
    return self._status

  # Audio API
  def send_audio(self, audio: bytes) -> None:
    if self._status in ("ending", "ended"):
      return
    with self._state_lock:
      self._audio_buffer += audio
      ws = self._ws
      is_open = bool(ws and ws.ready_state == WS_STATES.OPEN)
    if is_open and ws:
      with contextlib.suppress(Exception):
        ws.send(audio)

  def stop_recording(self) -> None:
    if self._status in ("ending", "ended"):
      return
    self._status = "ending"

    _ = self._event_emitter.emit("ending", LiveV2EndingMessage(code=1000))
    with self._state_lock:
      ws = self._ws
      is_open = bool(ws and ws.ready_state == WS_STATES.OPEN)
      if is_open and ws:
        with contextlib.suppress(Exception):
          ws.send(json.dumps({"type": "stop_recording"}))
        self._pending_stop = False
      else:
        self._pending_stop = True

  def end_session(self) -> None:
    self._do_destroy(1000, "Session ended by user")

  # Internals
  def _init_session(self) -> LiveV2InitResponse:
    try:
      options = with_acknowledgments_enabled(self._options)
      resp = self._http_client.post("/v2/live", json=options.to_dict())
      return LiveV2InitResponse.from_json(resp.content)
    except Exception as err:
      _ = self._event_emitter.emit("error", err)
      self._do_destroy(1006, "Couldn't start a new session")
      raise err

  def _start_session(self) -> None:
    try:
      session = self._init_session_response
      if not session:
        return

      self._status = emit_started_if_needed(self._event_emitter, self._status, session)
      maybe_emit_start_session_message(self._event_emitter, self._options, session)
      self._connect_ws(session.url)
    except Exception as err:
      _ = self._event_emitter.emit("error", err)
      self._do_destroy(1006, "Couldn't start a new session")

  def _connect_ws(self, ws_url: str) -> None:
    ws = self._ws_client.create_session(ws_url)
    self._ws = ws

    def _on_connecting(info: dict[str, Any]) -> None:
      self._status = "connecting"
      attempt = int(info.get("attempt", 1))
      _ = self._event_emitter.emit("connecting", LiveV2ConnectingMessage(attempt=attempt))

    def _on_open(info: dict[str, Any]) -> None:
      # Flush any buffered audio through the worker queue
      with self._state_lock:
        buffered = self._audio_buffer
        pending_stop = self._pending_stop
      if buffered and len(buffered):
        with contextlib.suppress(Exception):
          send_audio_in_chunks(ws, buffered)
      if pending_stop:
        with self._state_lock:
          self._pending_stop = False
        with contextlib.suppress(Exception):
          ws.send(json.dumps({"type": "stop_recording"}))

      if self._status == "ending":
        # Stop message already enqueued above if needed
        return

      self._status = "connected"
      attempt = int(info.get("attempt", 1))
      _ = self._event_emitter.emit("connected", LiveV2ConnectedMessage(attempt=attempt))
      self._ready_event.set()

    def _on_message(evt: dict[str, Any]) -> None:
      raw = evt.get("data")
      try:
        message = parse_ws_message(raw)
      except Exception as parse_err:
        _ = self._event_emitter.emit("error", parse_err)
        return

      if should_emit_ws_message(message, self._options.messages_config):
        _ = self._event_emitter.emit("message", message)

      if getattr(message, "type", None) == "audio_chunk":
        data = getattr(message, "data", None)
        if getattr(message, "acknowledged", False) and data:
          with self._state_lock:
            byte_end = int((data.byte_range)[1])
            self._audio_buffer, self._bytes_sent = trim_acknowledged_audio_buffer(
              self._audio_buffer,
              self._bytes_sent,
              byte_end,
            )

    def _on_error(err: Exception) -> None:
      _ = self._event_emitter.emit("error", err)

    def _on_close(evt: dict[str, Any]) -> None:
      code = int(evt.get("code", 1000) or 1000)
      reason = evt.get("reason")
      self._ws_stop.set()
      self._do_destroy(code, reason)

    ws.onconnecting = _on_connecting
    ws.onopen = _on_open
    ws.onmessage = _on_message
    ws.onerror = _on_error
    ws.onclose = _on_close
    # Start background receive/retry loop
    with contextlib.suppress(Exception):
      ws.start()

  def _lifecycle_worker(self) -> None:
    try:
      if self._existing_session:
        self._init_session_response = self._existing_session
      else:
        self._init_session_response = self._init_session()
      self._start_session()
      # 3) Keep thread alive until stop/end to let join() wait for session end
      while not self._ws_stop.is_set() and self._status != "ended":
        time.sleep(0.05)
    except Exception as err:
      _ = self._event_emitter.emit("error", err)
      self._do_destroy(1006, "Worker aborted due to error")
    finally:
      # Nothing to cleanup here; _do_destroy handles ws closure and listeners
      pass

  def _do_destroy(self, code: int = 1006, reason: str | None = None) -> None:
    if self._status == "ended":
      return

    self._status = emit_session_ending_events(self._event_emitter, self._status, code, reason)
    # Signal worker to stop
    self._ws_stop.set()
    # Close ws
    ws = self._ws
    self._ws = None
    if ws:
      with contextlib.suppress(Exception):
        ws.close(code=1001, reason="Aborted")

    # Clear buffers & listeners
    self._audio_buffer = bytes([])
    self._event_emitter.remove_all_listeners()

  # Threading helpers
  def wait_until_ready(self, timeout: float | None = None) -> bool:
    """Block until WebSocket is connected (ready), or timeout elapses.

    Returns True if ready, False if timeout.
    """
    return self._ready_event.wait(timeout)

  def join(self, timeout: float | None = None) -> bool:
    """Wait for the background worker thread to exit.

    Returns True if the worker is no longer alive after the join attempt.
    """
    t = self._ws_thread
    if not t:
      return True
    if t is threading.current_thread():
      # Avoid deadlocking by joining self
      return not t.is_alive()
    t.join(timeout)
    return not t.is_alive()
